src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ImmutableDispatchRuntime.cs

```diff
@@ -640,6 +667,44 @@ internal async Task<MessageRpc> ProcessMessageAsync(MessageRpc rpc)
             return rpc;
         }
 
+        private (bool isConfigured, HttpContext httpContext, string authenticationScheme) IsCustomAuthenticationSchemeConfigured()
+            =>  OperationContext.Current.IncomingMessageProperties.TryGetValue<HttpContext>(
+                    "Microsoft.AspNetCore.Http.HttpContext", out var httpContext)
+                && httpContext.Items.TryGetValue("CoreWCF.Channels.HttpTransportSettings.CustomAuthenticationScheme",
+                    out var authenticationScheme)
+                && authenticationScheme is string authenticationSchemeValue
+                && !string.IsNullOrEmpty(authenticationSchemeValue)
+                ? (true, httpContext, authenticationSchemeValue)
+                : (false, null, null);
+
+        private async ValueTask<(bool isAuthorized, Message message)> EvaluateAuthorizationPolicies(MessageRpc rpc)
+        {
+            var httpContext = OperationContext.Current.IncomingMessageProperties["Microsoft.AspNetCore.Http.HttpContext"] as HttpContext;
+            string authenticationScheme = httpContext.Items["CoreWCF.Channels.HttpTransportSettings.CustomAuthenticationScheme"] as string;
+            var principal = httpContext.User;
```

Comment:
These two lines will throw on non-http stacks. What we should do is make the principal available in the request message properties. Message.Properties.Security.ServiceSecurityContext contains a ServiceSecurityContext. This class has an AuthorizationContext which has a Properties class where you can add arbitrary named objects. If you look at ServiceSecurityContext.GetIdentities, you can see how we use this to store an `IList<IIdentity>` in the authorization context as AuthorizationContext doesn't directly support exposing IIdentity objects. We could to a similar thing to store the ClaimsPrincipal from HttpContext.User in the Http code. Then when we add later add support for non-http transports to use the authentication and policy mechanisms from asp.net core in a general purpose way, we can store the ClaimsPrincipal in the same way and this code will automatically light up.

Reply:
I dropped this code and created a property named `ClaimsPrincipal` on `ServiceSecurityContext`. This property returns the `ClaimsPrincipal` from `HttpContext.User` that has been previsously stored in `AuthorizationContext.Properties`.